src/Aspire.Hosting/Extensions/ProjectResourceBuilderExtensions.cs

```diff
@@ -44,18 +44,139 @@ public static IResourceBuilder<ProjectResource> AddProject(this IDistributedAppl
+
+        projectPath = PathNormalizer.NormalizePathForCurrentPlatform(Path.Combine(builder.AppHostDirectory, projectPath));
+
+        return builder.AddResource(project)
+                      .WithAnnotation(new ProjectMetadata(projectPath))
+                      .WithProjectDefaults(excludeLaunchProfile: launchProfileName is null, launchProfileName);
+    }
+
+    private static IResourceBuilder<ProjectResource> WithProjectDefaults(this IResourceBuilder<ProjectResource> builder, bool excludeLaunchProfile, string? launchProfileName)
     {
         // We only want to turn these on for .NET projects, ConfigureOtlpEnvironment works for any resource type that
         // implements IDistributedApplicationResourceWithEnvironment.
         builder.WithEnvironment("OTEL_DOTNET_EXPERIMENTAL_OTLP_EMIT_EXCEPTION_LOG_ATTRIBUTES", "true");
         builder.WithEnvironment("OTEL_DOTNET_EXPERIMENTAL_OTLP_EMIT_EVENT_LOG_ATTRIBUTES", "true");
         builder.WithOtlpExporter();
         builder.ConfigureConsoleLogs();
+
+        if (excludeLaunchProfile)
+        {
+            builder.WithAnnotation(new ExcludeLaunchProfileAnnotation());
+            return builder;
+        }
+
+        if (!string.IsNullOrEmpty(launchProfileName))
+        {
+            builder.WithAnnotation(new LaunchProfileAnnotation(launchProfileName));
+        }
+
+        var projectResource = builder.Resource;
+
+        if (builder.ApplicationBuilder.ExecutionContext.IsRunMode)
+        {
+            // Automatically add EndpointAnnotation to project resources based on ApplicationUrl set in the launch profile.
+            var launchProfile = projectResource.GetEffectiveLaunchProfile();
+            if (launchProfile is null)
+            {
+                return builder;
+            }
+
+            var urlsFromApplicationUrl = launchProfile.ApplicationUrl?.Split(';') ?? [];
+            foreach (var url in urlsFromApplicationUrl)
+            {
+                var uri = new Uri(url);
+
+                var endpointAnnotations = projectResource.Annotations.OfType<EndpointAnnotation>().Where(sb => string.Equals(sb.Name, uri.Scheme, StringComparisons.EndpointAnnotationName));
+                if (endpointAnnotations.Any(sb => sb.IsProxied))
+                {
+                    // If someone uses WithEndpoint in the dev host to register a endpoint with the name
+                    // http or https this exception will be thrown.
+                    throw new DistributedApplicationException($"Endpoint with name '{uri.Scheme}' already exists.");
+                }
+
+                if (endpointAnnotations.Any())
+                {
+                    // We have a non-proxied endpoint with the same name as the 'url', don't add another endpoint for the same name
+                    continue;
+                }
+
+                // TODO:
+                //int? port = options.Value.RandomizePorts is true ? null : uri.Port;
```

Comment:
This is the only difference from copying this code over

Reply:
This is new code @ReubenBond added. There's likely a way to flow this through via the service provider (even if it looks gross 😄)

Reply:
@ReubenBond why not flow this to DCP directly and magically handle it so it's not done here?

Reply:
How much magic do you want here versus in DCP? This setting just tells the app model to not explicitly flow a port, so DCP can assign it. If we flow this to DCP, we'll be telling DCP that we have a port but that it shouldn't assign it, which is weird to to me

Reply:
This is literally a DCP setting isn't it? (it's on DCP options)

Reply:
It an option for when we're running via DCP, but not option for DCP itself 😅